File: 0001-1000-easy-python/0225-implement-stack-using-queues-v1.py
# https://leetcode.com/problems/implement-stack-using-queues
from collections import deque

class MyStack(object):

    def __init__(self):
        self.queue = deque()
        self.front = None

    # O(n)
    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        if (self.front != None):
            self.queue.appendleft(self.front)
            for i in range(len(self.queue) - 1):
                self.queue.appendleft(self.queue.pop())
        self.front = x

    # O(1)
    def pop(self):
        """
        :rtype: int
        """
        tmp = self.front
        if len(self.queue) == 0:
            self.front = None
        else:
            self.front = self.queue.pop()
        return tmp

    # O(1)
    def top(self):
        """
        :rtype: int
        """
        return self.front

    # O(1)
    def empty(self):
        """
        :rtype: bool
        """
        return len(self.queue) == 0 and self.front == None

# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: 0001-1000-easy-cpp/0225-implement-stack-using-queues-v1.cpp
// https://leetcode.com/problems/implement-stack-using-queues
class MyStack {
private:
    std::queue<int> queue;

public:
    MyStack() {
    }

    // O(n)
    void push(int x) {
        this->queue.push(x);
        for (int i = 0; i < this->queue.size() - 1; i++) {
            this->queue.push(this->queue.front());
            this->queue.pop();
        }
    }

    // O(1)
    int pop() {
        int tmp = this->queue.front();
        this->queue.pop();
        return tmp;
    }

    // O(1)
    int top() {
        return this->queue.front();
    }

    // O(1)
    bool empty() {
        return this->queue.empty();
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */

